File: app/src/main/java/com/adrzdv/mtocp/ui/viewmodel/model/ViolationViewModel.java
package com.adrzdv.mtocp.ui.viewmodel.model;

import static com.adrzdv.mtocp.domain.model.enums.RevisionType.ALL;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.adrzdv.mtocp.data.db.entity.ViolationEntity;
import com.adrzdv.mtocp.domain.model.enums.RevisionType;
import com.adrzdv.mtocp.domain.repository.ViolationRepository;
import com.adrzdv.mtocp.mapper.ViolationMapper;
import com.adrzdv.mtocp.ui.model.ViolationDto;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ViolationViewModel extends ViewModel {

    private final ViolationRepository repository;
    private final MutableLiveData<List<ViolationDto>> filteredViolations;
    private List<ViolationEntity> allViolations;
    private String currentSearchString;
    private RevisionType currentRevisionType;
    private final ExecutorService executor;


    public ViolationViewModel(ViolationRepository repository) {
        this.repository = repository;
        this.filteredViolations = new MutableLiveData<>(new ArrayList<>());
        this.allViolations = new ArrayList<>();
        this.currentSearchString = "";
        this.currentRevisionType = ALL;
        this.executor = Executors.newSingleThreadExecutor();
        loadViolations();
    }

    private void loadViolations() {

        executor.execute(() -> {
            allViolations = repository.getAll();
            applyFilters();
        });
    }

    public LiveData<List<ViolationDto>> getFilteredViolations() {
        return filteredViolations;
    }

    public void filterDataByString(String str) {
        currentSearchString = str != null ? str.trim() : "";
        applyFilters();
    }

    public void filterDataByRevisionType(RevisionType revType) {
        currentRevisionType = revType != null ? revType : ALL;
        applyFilters();
    }

    private void applyFilters() {
        List<ViolationEntity> result = new ArrayList<>(allViolations);

        result = switch (currentRevisionType) {
            case IN_TRANSIT -> result.stream().filter(ViolationEntity::getInTransit).toList();
            case AT_START_POINT ->
                    result.stream().filter(ViolationEntity::getAtStartPoint).toList();
            case AT_TURNROUND_POINT ->
                    result.stream().filter(ViolationEntity::getAtTurnroundPoint).toList();
            case AT_TICKET_OFFICE ->
                    result.stream().filter(ViolationEntity::getAtTicketOffice).toList();
            default -> result;
        };

        String str = currentSearchString.toLowerCase();
        if (!str.isEmpty()) {
            result = result.stream().filter(violation -> {
                String name = violation.getName().toLowerCase();
                String code = String.valueOf(violation.getCode());
                return name.contains(str) || code.contains(str);
            }).toList();
        }

        List<ViolationDto> dtoList = result.stream()
                .map(ViolationMapper::fromEntityToDto)
                .toList();

        filteredViolations.postValue(dtoList);
    }

}

File: app/src/main/java/com/adrzdv/mtocp/ui/viewmodel/model/ViolationViewModel.kt
package com.adrzdv.mtocp.ui.viewmodel.model

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.adrzdv.mtocp.data.db.entity.ViolationFullInfo
import com.adrzdv.mtocp.data.repository.refcache.CacheRepository
import com.adrzdv.mtocp.domain.model.enums.RevisionType
import com.adrzdv.mtocp.domain.model.violation.ViolationDomain
import com.adrzdv.mtocp.mapper.toDomain
import com.adrzdv.mtocp.mapper.toUi
import com.adrzdv.mtocp.ui.model.dto.ViolationUi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.mapLatest
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

class ViolationViewModel(
    private val cacheRepository: CacheRepository
) : ViewModel() {

    private val _query = MutableStateFlow("")
    val query: StateFlow<String> = _query

    private val _revisionType = MutableStateFlow<RevisionType?>(null)
    val revisionType: StateFlow<RevisionType?> = _revisionType

    private val allEntities: List<ViolationFullInfo> =
        cacheRepository.violationCache.getAll()

    private val _items = MutableStateFlow<List<ViolationFullInfo>>(emptyList())
    val items: StateFlow<List<ViolationFullInfo>> = _items

    val suggestions: StateFlow<List<ViolationUi>> =
        _items
            .map { list -> list.map { it.toUi() } }
            .stateIn(
                viewModelScope,
                SharingStarted.WhileSubscribed(5000),
                emptyList()
            )

    init {
        observeFilters()
    }

    fun setRevisionType(type: RevisionType) {
        _revisionType.value = type
    }

    fun onQueryChange(str: String) {
        _query.value = str
    }

    fun resetQuery() {
        _query.value = ""
    }

    fun getDomainByCode(code: Int): ViolationDomain {
        return _items.value.first { it.violation.code == code }.toDomain()
    }

    @OptIn(FlowPreview::class, ExperimentalCoroutinesApi::class)
    private fun observeFilters() {

        viewModelScope.launch {
            combine(
                _query.debounce(300),
                _revisionType
            ) { query, type -> query to type }
                .mapLatest { (query, type) ->
                    allEntities
                        .asSequence()
                        .filter { info ->
                            type == null ||
                                    type == RevisionType.ALL ||
                                    info.revisionTypes.any { revisionTypeEntity ->
                                revisionTypeEntity.name.equals(
                                    type.revisionTypeTitle,
                                    ignoreCase = true
                                )
                            }
                        }
                        .filter { info ->
                            query.isBlank() ||
                                    info.violation.code == query.toIntOrNull() ||
                                    info.violation.description.contains(query, ignoreCase = true) ||
                                    info.violation.criteria.contains(query, ignoreCase = true) ||
                                    info.violation.measure.contains(query, ignoreCase = true) ||
                                    info.departments.any {
                                        it.shortName.contains(query, ignoreCase = true) ||
                                                it.fullName.contains(query, ignoreCase = true)
                                    } ||
                                    info.divisions.any {
                                        it.shortName.contains(query, ignoreCase = true) ||
                                                it.name.contains(query, ignoreCase = true)
                                    } ||
                                    info.revisionTypes.any {
                                        it.name.contains(query, ignoreCase = true)
                                    }
                        }
                        .toList()
                }
                .flowOn(Dispatchers.Default)
                .collect { filtered ->
                    _items.value = filtered
                }
        }
    }
}
